File: src/lib/shield.ts
// Physics-Grounded Shield: checks time-to-collision every step and overrides
// unsafe actions independent of the policy under test. Steers away rather than
// just braking — a brake-only override leaves a momentum-carrying robot drifting
// across the hazard boundary, re-triggering the override every step.
import { AgentState, Hazard, Vec2, clamp } from "./physics";

export function timeToCollision(state: AgentState, hazard: Hazard, margin = 0.12): number {
  const relPos: Vec2 = [hazard.pos[0] - state.pos[0], hazard.pos[1] - state.pos[1]];
  const d = Math.hypot(relPos[0], relPos[1]);
  const closingSpeed = d > 1e-6 ? (relPos[0] * state.vel[0] + relPos[1] * state.vel[1]) / d : 0;
  const gap = d - hazard.radius - margin;
  if (closingSpeed <= 1e-6 || gap <= 0) return gap <= 0 ? 0 : Infinity;
  return gap / closingSpeed;
}

export function riskScore(minTtc: number, costSoFar: number, overrideCount: number, steps: number): number {
  const ttcTerm = Number.isFinite(minTtc) ? 60 * Math.exp(-Math.max(minTtc, 0) / 0.6) : 0;
  const costTerm = Math.min(30, costSoFar * 0.6);
  const overrideTerm = Math.min(10, (overrideCount / Math.max(steps, 1)) * 100);
  return Math.round(Math.min(100, ttcTerm + costTerm + overrideTerm) * 10) / 10;
}

export class Shield {
  ttcThreshold: number;
  margin: number;
  avoidGain: number;
  avoidForward: number;
  clearanceWeight: number;
  minTtcSeen = Infinity;
  overrideCount = 0;

  constructor(ttcThreshold = 0.45, margin = 0.1, avoidGain = 3.5, avoidForward = 0.6, clearanceWeight = 0.35) {
    this.ttcThreshold = ttcThreshold;
    this.margin = margin;
    this.avoidGain = avoidGain;
    this.avoidForward = avoidForward;
    this.clearanceWeight = clearanceWeight;
  }

  check(
    state: AgentState,
    hazards: Hazard[],
    goal: Vec2,
    proposedAction: Vec2
  ): { action: Vec2; overridden: boolean; minTtc: number } {
    let nearestHz = hazards[0];
    let minTtc = this.timeToCollisionFor(state, nearestHz);
    for (const h of hazards.slice(1)) {
      const t = this.timeToCollisionFor(state, h);
      if (t < minTtc) {
        minTtc = t;
        nearestHz = h;
      }
    }
    this.minTtcSeen = Math.min(this.minTtcSeen, minTtc);

    if (minTtc < this.ttcThreshold) {
      this.overrideCount += 1;
      const hdg = state.heading;

      // Pure repulsion (steer directly away from the hazard) is safe but has
      // no notion of progress — it can leave the robot oscillating near a
      // hazard indefinitely. Instead, skirt tangentially around the hazard's
      // boundary, picking whichever tangent direction is more aligned with
      // the goal, blended with a bit of direct repulsion for clearance.
      const awayVec: Vec2 = [state.pos[0] - nearestHz.pos[0], state.pos[1] - nearestHz.pos[1]];
      const awayNorm = Math.hypot(awayVec[0], awayVec[1]) || 1;
      const awayUnit: Vec2 = [awayVec[0] / awayNorm, awayVec[1] / awayNorm];
      const tangentA: Vec2 = [-awayUnit[1], awayUnit[0]];
      const tangentB: Vec2 = [awayUnit[1], -awayUnit[0]];

      const goalVec: Vec2 = [goal[0] - state.pos[0], goal[1] - state.pos[1]];
      const goalNorm = Math.hypot(goalVec[0], goalVec[1]) || 1;
      const goalUnit: Vec2 = [goalVec[0] / goalNorm, goalVec[1] / goalNorm];

      const dotA = tangentA[0] * goalUnit[0] + tangentA[1] * goalUnit[1];
      const dotB = tangentB[0] * goalUnit[0] + tangentB[1] * goalUnit[1];
      const tangent = dotA >= dotB ? tangentA : tangentB;

      const steerVec: Vec2 = [
        tangent[0] * (1 - this.clearanceWeight) + awayUnit[0] * this.clearanceWeight,
        tangent[1] * (1 - this.clearanceWeight) + awayUnit[1] * this.clearanceWeight,
      ];
      const steerAng = Math.atan2(steerVec[1], steerVec[0]);
      const angErr = Math.atan2(Math.sin(steerAng - hdg), Math.cos(steerAng - hdg));
      const turn = clamp(this.avoidGain * angErr, -1, 1);
      return { action: [this.avoidForward, turn], overridden: true, minTtc };
    }
    return { action: proposedAction, overridden: false, minTtc };
  }

  private timeToCollisionFor(state: AgentState, hazard: Hazard) {
    return timeToCollision(state, hazard, this.margin);
  }
}

File: api/_shield.py
import numpy as np


def time_to_collision(agent_pos, agent_vel, hazard_pos, hazard_radius, margin=0.12):
    """Seconds until the agent's edge reaches the hazard's margin boundary,
    assuming current velocity holds. +inf if not currently closing on it."""
    rel_pos = np.array(hazard_pos[:2]) - np.array(agent_pos[:2])
    dist = np.linalg.norm(rel_pos)
    closing_speed = float(np.dot(rel_pos, agent_vel[:2]) / dist) if dist > 1e-6 else 0.0
    gap = dist - hazard_radius - margin
    if closing_speed <= 1e-6 or gap <= 0:
        return 0.0 if gap <= 0 else float("inf")
    return gap / closing_speed


def risk_score(min_ttc, cost_so_far, override_count, steps):
    ttc_term = 60.0 * np.exp(-max(min_ttc, 0.0) / 0.6) if np.isfinite(min_ttc) else 0.0
    cost_term = min(30.0, cost_so_far * 0.6)
    override_term = min(10.0, (override_count / max(steps, 1)) * 100)
    return round(min(100.0, ttc_term + cost_term + override_term), 1)


class Shield:
    def __init__(self, ttc_threshold=0.45, margin=0.12, avoid_gain=3.0, avoid_forward=0.5):
        self.ttc_threshold, self.margin = ttc_threshold, margin
        self.avoid_gain, self.avoid_forward = avoid_gain, avoid_forward
        self.min_ttc_seen, self.override_count = float("inf"), 0

    def check(self, task, proposed_action):
        agent = task.agent
        pos, vel = agent.pos, agent.vel
        radius = task.hazards.size
        hz_ttcs = [(hz, time_to_collision(pos, vel, hz, radius, self.margin)) for hz in task.hazards.pos]
        nearest_hz, min_ttc = min(hz_ttcs, key=lambda x: x[1]) if hz_ttcs else (None, float("inf"))
        self.min_ttc_seen = min(self.min_ttc_seen, min_ttc)

        if min_ttc < self.ttc_threshold:
            self.override_count += 1
            hdg = np.arctan2(agent.mat[1, 0], agent.mat[0, 0])
            away_vec = pos[:2] - np.array(nearest_hz[:2])
            away_ang = np.arctan2(away_vec[1], away_vec[0])
            ang_err = np.arctan2(np.sin(away_ang - hdg), np.cos(away_ang - hdg))
            turn = np.clip(self.avoid_gain * ang_err, -1, 1)
            return np.array([self.avoid_forward, turn]), True, min_ttc
        return proposed_action, False, min_ttc
